util: Replace boost sleep with std sleep

maflcko opened the pull request:
We don't use the interruptible feature of boost's sleep anywhere, so replace it with the sleep in `std::thread`

laanwj commented on src/util/time.cpp:
```diff
@@ -53,41 +59,22 @@ int64_t GetMockTime()
- * until fixed in 1.52. Use the deprecated sleep method for the broken case.
- * See: https://svn.boost.org/trac/boost/ticket/7238
- */
-#if defined(HAVE_WORKING_BOOST_SLEEP_FOR)
```
oh wow, can we finally get rid of this abomination?
feel free to delete the detection of this as well from the build system!

MarcoFalke replied:
Good idea! Done.

MarcoFalke commented:
|                  | before | after  |
|------------------|--------|--------|
| compile time (s) | 8.63   | 2.61   |
| compile rss (kB) | 446368 | 101168 |
| run time (ns)    | 85     | 25     |

Source:

```
make
./src/bench/bench_bitcoin --filter=BenchTime.*
```

MarcoFalke commented:
Removed unused configure checks

sipa commented:
Are the std:: versions compatible with our use of boost threadgroup interrupts?

laanwj commented:
> Are the std:: versions compatible with our use of boost threadgroup interrupts?

Oh crap I knew there was a reason we hadn't got rid of the boost::sleep madness yet :cry:

practicalswift commented:
> Hopefully that has no observable effects other than improved code clarity, faster compilation and faster run time.

There is an very observable effect of this PR on 32-bit systems:

Prior to this PR `bitcoind` would shut down with an assertion failure when the `time_t` wraparound occurs on 32-bit systems.

After this PR `bitcoind` will keep running but with negative time returned from these functions :-)

practicalswift commented:
Perhaps we should add a test making sure `bitcoind` on 32-bit systems behave the way we want it to -- which I assume is immediate shutdown -- at time of the 32-bit `time_t` wraparound event.

laanwj commented:
How do other programs handle this? I though Linux switched to 64 bit timer values on 32-bit systems internally a while ago, but I see on ARM `time_t` is still 32 bit.

MarcoFalke commented:
> After this PR bitcoind will keep running but with negative time returned from these functions :-)

Can you please explain a bit more why this happens. Be reminded that the predefined `std::chrono::duration` use an underlying integer type that is large enough (for the y2038 to be of no problem).

practicalswift commented:
> > After this PR bitcoind will keep running but with negative time returned from these functions :-)
> 
> Can you please explain a bit more why this happens. Be reminded that the predefined `std::chrono::duration`s use an underlying integer type that is large enough (for year 2038 to be of no problem).

As long as the syscall used returns 32-bit time on 32-bit systems this cannot be solved no matter what integer types are used in the standard library, no?

MarcoFalke commented:
Yeah, and the you can't run the underlying system anyway. There is no need to try to figure out how to make Bitcoin Core work on a broken system.

practicalswift commented:
@MarcoFalke Perhaps we should start thinking about a deprecation plan for 32-bit systems (see #16096)? Even if this is 19 years away software tends to live longer than originally anticipated :-)

Commit:
util: Add UnintrruptibleSleep

Commit:
scripted-diff: Replace MilliSleep with UninterruptibleSleep

This is safe because MilliSleep is never executed in a boost::thread,
the only type of thread that is interruptible.

* The RPC server uses std::thread
* The wallet is either executed in an RPC thread or the main thread
* bitcoin-cli, benchmarks and tests are only one thread (the main thread)

-BEGIN VERIFY SCRIPT-
sed -i --regexp-extended -e 's/MilliSleep\((\S+)\);/UninterruptibleSleep(std::chrono::milliseconds{\1});/g' $(git grep -l MilliSleep)
-END VERIFY SCRIPT-

Commit:
util: Remove unused MilliSleep

MarcoFalke commented:
I've removed the changes on the time getters, so that this doesn't drift off into a discussion about the y2038 problem

practicalswift commented:
ACK fae86c38bca5c960462e53975314a0749db5d17d -- patch looks correct

sipa commented:
Concept and code review ACK fae86c38bca5c960462e53975314a0749db5d17d

fanquake reviewed (approved):
ACK fae86c38bca5c960462e53975314a0749db5d17d - note that an instance of `DHAVE_WORKING_BOOST_SLEEP_FOR` was missed in the [linter](https://github.com/bitcoin/bitcoin/blob/master/test/lint/extended-lint-cppcheck.sh#L69), but that can be cleaned up later.

ajtowns commented:
ACK fae86c38bca5c960462e53975314a0749db5d17d quick code review

laanwj commented:
Postmortem ACK.
